### PR TITLE
Send Image As Sticker

### PR DESCRIPTION
**Pull request by andersondeoliveiramachado**

### CODE REVIEW

**Comment by smashah** [comment]
This only works when the base image is already a webp image.

TODO: Implement on-the-fly webp conversion:
https://github.com/scionoftech/webp-converter

**Comment by andersondeoliveiramachado** [comment]
> This only works when the base image is already a webp image.
> 
> TODO: Implement on-the-fly webp conversion:
> https://github.com/scionoftech/webp-converter

https://sharp.pixelplumbing.com/api-output#webp

**Comment by smashah** [comment]
@andersondeoliveiramachado yeah that's a much better lib.

**Comment by andersondeoliveiramachado** [comment]
> @andersondeoliveiramachado yeah that's a much better lib.

To use it in the case it would have to be in the sulla part and not in WAPI, correct?

**Comment by smashah** [comment]
@andersondeoliveiramachado yes. I'm working on it. Should be ready in a bit. Standby

**Comment by smashah** [comment]
It's ready and working @andersondeoliveiramachado 

1. Run `npm i`
2. Start the demo
3. Send a picture to the host phone.

File: src/controllers/browser.ts [hunk 1]
```diff
@@ -63,6 +63,9 @@ export async function initWhatsapp(sessionId?: string, puppeteerConfigOverride?:
 export async function injectApi(page: Page) {
   // const preloadFile = fs.readFileSync('./preload', 'utf8');
   // await page.evaluateOnNewDocument(preloadFile);
+  await page.addScriptTag({
+    path: require.resolve(path.join(__dirname, '../lib', 'sha256.js'))
```
Comment (smashah) [inline src/controllers/browser.ts]: If this is required for a feautre of WAPI.js itself to work then it should be a oneliner in there already.

File: src/lib/sticker.js [hunk 2]
```diff
@@ -1,3 +1,8 @@
+// SEND IMAGEM AS STICKER
```
Comment (smashah) [inline src/lib/sticker.js]: TODO: delete file once this feature is completed

File: src/lib/wapi.js [hunk 3]
```diff
@@ -2216,29 +2219,101 @@ window.WAPI.demoteParticipant = function (idGroup, idParticipant, done) {
  * Send Sticker
  * @param {*} sticker
```
Comment (smashah) [inline src/lib/wapi.js]: requires documentation

File: src/lib/wapi.js [hunk 4]
```diff
@@ -2216,29 +2219,101 @@ window.WAPI.demoteParticipant = function (idGroup, idParticipant, done) {
...
-		// stick.height = 512;
-		// stick.width = 512;
-		await sticker.initialize();
-		await sticker.sendToChat(chat);
-		if (done !== undefined) done(true);
-	})
-	.catch(e => {
-		if (done !== undefined) done(false);
-	});
+window.WAPI.sendSticker = async function ({sticker, chatid}) {
```
Comment (smashah) [inline src/lib/wapi.js]: This should probably be a private function

File: src/lib/wapi.js [hunk 5]
```diff
@@ -2216,29 +2219,101 @@ window.WAPI.demoteParticipant = function (idGroup, idParticipant, done) {
...
+    var result = '';
+    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    var charactersLength = characters.length;
+    for ( var i = 0; i < length; i++ ) {
+        result += characters.charAt(Math.floor(Math.random() * charactersLength));
+    }
+    return result;
+};
+
+window.WAPI.encryptAndUploadFile = async function ({ type, data }) {
```
Comment (smashah) [inline src/lib/wapi.js]: Needs to use params instead of object to maintain consistency throughout WAPI.js

File: src/lib/wapi.js [hunk 6]
```diff
@@ -2216,29 +2219,101 @@ window.WAPI.demoteParticipant = function (idGroup, idParticipant, done) {
...
+window.WAPI.sendSticker = async function ({sticker, chatid}) {
+    // console.log(sticker);
+    // console.log(chatid);
+    var chat = Store.Chat.get(chatid)
+    console.log(chat);
+	// var idUser = new window.Store.UserConstructor(chatid, { intentionallyUsePrivateConstructor: true });
+	// return  (chat) => {
+        // ephemeralStartTimestamp
+		let stick = new window.Store.Sticker.modelClass();		
+		stick.__x_clientUrl = sticker.clientUrl;
```
Comment (smashah) [inline src/lib/wapi.js]: Check: can this be spread/Object.assign'ed?

Comment (smashah) [reply src/lib/wapi.js]: No

File: src/lib/wapi.js (hunk 4 above)
Comment (smashah) [reply src/lib/wapi.js]: 👍

File: src/lib/wapi.js (hunk 5 above)
Comment (smashah) [reply src/lib/wapi.js]: 👍

**Comment by smashah** [comment]
@github-actions run

<details>
<summary>⚡ Release!  ⚡</summary>

```js
(async () => {
function exec(cmd) {
  console.log(execSync(cmd).toString());
}

// Config
const gitUserEmail = "github-actions[bot]@users.noreply.github.com";
const gitUserName = "github-actions[bot]";

exec(`echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > .npmrc`);
exec(`git config --global user.email "${gitUserEmail}"`);
exec(`git config --global user.name "${gitUserName}"`);
exec(`npm i`);
exec(`npm run release-ci`);

var result = execSync(`npx auto-changelog -o ./tempchangelog.txt --commit-limit false --template ./compact-keepachangelog.hbs --stdout`).toString();

    await postComment(result);
  })();
```
</details>

**Comment by smashah** [comment]
# Changelog

## 🚀 Release [1.8.17](https://github.com/smashah/sulla/compare/1.8.16...1.8.17) (2020-03-11)

- Send Image As Sticker implemented [`#202`](https://github.com/smashah/sulla/pull/202)
- Release 1.8.17 [`c9f6222`](https://github.com/smashah/sulla/commit/c9f6222c1c746610a5304707eed6779bcf304f4b)
- transferred jsSha to wapi, resolves review comments [`3b0178e`](https://github.com/smashah/sulla/commit/3b0178e7e21e58cc6859d957ac9f35069e8095e7)
- added metadata based height and width [`c2660c2`](https://github.com/smashah/sulla/commit/c2660c2593ec3880e80244e008845b6b4809c969)
- deleted sticker.js [`d4d0712`](https://github.com/smashah/sulla/commit/d4d071234a4616b275fdf87bb77911218dcd728b)
- properly crop images [`b627cc8`](https://github.com/smashah/sulla/commit/b627cc838925247f3ff96960966b4ef81ae9f795)
- remove browser import for sha256.js [`eb7f339`](https://github.com/smashah/sulla/commit/eb7f339876ddae76330844733eb6083b3e55a98b)
- added documentation for sendImageAsSticker [`f85258b`](https://github.com/smashah/sulla/commit/f85258b8bd9590e7a77ab29fb122e619daad85ed)
- The final  ✅ [`dfd6ff7`](https://github.com/smashah/sulla/commit/dfd6ff7d1eb48ac2a88ac6055b2c8601e51a39b0)

**Comment by smashah** [comment]
Just in case anyone stumbles across this PR, this is how you use [`sendImageAsSticker`](https://open-wa.github.io/wa-automate-nodejs/classes/whatsapp.html#sendimageassticker)

```javascript
      const mimetype = 'jpg'
      const b64 = fs.readFileSync('./path/to-file.jpg').toString('base64');
      await client.sendImageAsSticker(`data:${mimetype};base64,${b64}`, '0000000000@c.us')
```

@renato-macedo

**Comment by renato-macedo** [comment]
Hello, I tried this but only the text message was sent and didn't throw any error

```javascript

const fs = require('fs')
const wa = require('@open-wa/wa-automate');

wa.create({
  headless: false,
  devtools: true 
}).then(client => start(client));

function start(client) {
  client.onMessage(async message => {
    if (message.body === 'Hi') {
      await client.sendText(message.from, '👋 Hello!')
      const mimetype = 'jpg'
      const b64 = fs.readFileSync('./assets/image.jpg').toString('base64')
      await client.sendImageAsSticker(`data:${mimetype};base64,${b64}`, message.from)
    }
  });
}
```

**Comment by smashah** [comment]
@renato-macedo can you share the image you are trying to use please

**Comment by renato-macedo** [comment]
@smashah I'm using my profile pic https://avatars2.githubusercontent.com/u/31698109?s=400&u=5788383d90807c8c157a778426431fff3d6dd8f3&v=4
I'll open a issue. Thank you for the patience.

**Comment by roelandp** [comment]
Fyi just in case you have not solved this (or for future folks reading this).... 

In your sample code in the replies above @smashah which was then/copied in his reply by @renato-macedo you guys actually flipped around the 2 parameters expected for the function:

according to the source it should be `sendImageAsSticker(to, imgdataB64)` and not the other way around :)

**Comment by NasirKhalid24** [comment]
Is there any way to disable the center cropping when sending image as sticker implemented in [b627cc8](https://github.com/open-wa/wa-automate-nodejs/pull/202/commits/b627cc838925247f3ff96960966b4ef81ae9f795)

**Pull request merged**